### PR TITLE
feat: add licence filter for PURLs (TC-2831)

### PR DESCRIPTION
**Pull request by mrizzi**
Adds license-based filtering to PURL queries, enabling searches like `license=GPLv3+`.

**Changes**

  - Implemented two-phase filtering strategy:
    - SPDX: Uses `expand_license_expression()` for LicenseRef resolution
    - CycloneDX: Direct text matching on license field
    - Results combined via UNION DISTINCT and used to filter main query to filter by `WHERE id IN (...)`
  - Added PostgreSQL function `expand_license_expression()` to resolve SPDX LicenseRef references
  - Created performance indexes on `licensing_infos` table
  - Added `get_constraint_for_field()` for extracting license constraints

Relates to [TC-2831](https://issues.redhat.com/browse/TC-2831)

## Summary by Sourcery

Add license filtering support to the PURL service by extending the query builder to recognize a license field, implementing two-phase subqueries for SPDX and CycloneDX documents, introducing a PostgreSQL function to resolve license references, and ensuring coverage with integration tests.

New Features:
- Support license-based filtering in PURL queries using a two-phase approach (SPDX with expand_license_expression and CycloneDX direct matching)

Enhancements:
- Add Query.get_constraint_for_field to extract field-specific constraints for license filtering

Build:
- Introduce a new database migration to create the expand_license_expression function and add an index on licensing_infos.sbom_id

Tests:
- Add integration tests for PURL license filtering covering positive, empty, malformed, and invalid scenarios

### CODE REVIEW

File: modules/fundamental/src/purl/service/mod.rs [hunk 1]
```diff
@@ -294,20 +297,53 @@ impl PurlService {
...
+            // 1. SPDX documents: Uses expand_license_expression() for LicenseRef resolution
+            // 2. CycloneDX documents: Direct text matching on license field
+            // The results are UNIONed and used to filter the main query.
+            let select_packages_filtering_by_license = select_packages_from_spdx
+                .union(UnionType::Distinct, select_packages_from_cyclonedx);
+
+            select = select.filter(
+                Condition::all().add(
+                    qualified_purl::Column::Id
+                        .in_subquery(select_packages_filtering_by_license.clone()),
```
Comment (bxf12315) [inline modules/fundamental/src/purl/service/mod.rs:341]: If I understand correctly, this subquery is used to obtain all purl IDs that meet the license condition, and then select the matching purls using the IN clause. So is it possible to pass the purl ID from the main query into the subquery, so that we can use EXISTS to determine whether the current purl contains the license condition?

**Comment by mrizzi** [comment]
> First, I would like to consider how to reuse these three methods: build_spdx_license_query, build_cyclonedx_license_query, and create_license_filtering_base_query. Second, I have done some testing, and the code is [here](https://github.com/mrizzi/trustify/pull/9). Currently, the search conditions are all for equality matches—I hope to add support for containment (i.e., 'contains') conditions. Third, I measured the execution performance in my tests, and it basically always takes more than 6 minutes.

1. Re the reuse, I think it's an important goal but this PR introduces those functions for the first time. Once merged, it will be up to next changes to reuse them if and how it will be possible. For the time being, it helps readability, avoiding never ending methods
2. In 239059f I've added a test with the `~` condition to prove it works as well
3. We know the `search_purls` scale test has unreliable results. We can work in analyzing the query execution plan and optimize it to ensure better performances. The `SbomIdIdx` index creation is meant to cover this: it supports the execution of
```
        SELECT license_id, name
        FROM licensing_infos
        WHERE sbom_id = sbom_id_param
```
which is the only query executed in the `expand_license_expression` PLSQL function.

File: migration/src/m0001140_expand_spdx_licenses_function/expand_up.sql [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+        WHERE sbom_id = sbom_id_param
+    LOOP
+        -- Replace the license_id with the license name in the expression
+        -- This handles exact matches and license references within expressions
+        result_text := regexp_replace(result_text, '\m' || license_mapping.license_id || '\M', license_mapping.name, 'g');
+END LOOP;
+
+RETURN result_text;
+END;
+$$ LANGUAGE plpgsql;
```
Comment (ctron) [inline migration/src/m0001140_expand_spdx_licenses_function/expand_up.sql:22]: IIRC, there had been several keywords like `parallel` which had a performance impact.

@JimFuller-RedHat can you help out here?

File: modules/fundamental/src/purl/service/mod.rs [hunk 3]
```diff
@@ -317,6 +353,61 @@ impl PurlService {
...
+
+    fn build_spdx_license_query(license_query: Query) -> Result<SelectStatement, Error> {
+        Ok(Self::create_license_filtering_base_query()
+            .filtering_with(
+                license_query,
+                Columns::default()
+                    .add_expr(
+                        "expanded_license",
+                        SimpleExpr::FunctionCall(
+                            Func::cust("expand_license_expression".into_identity())
```
Comment (ctron) [inline modules/fundamental/src/purl/service/mod.rs:378]: Bonus points for creating an `Ident` for this. Like: https://github.com/guacsec/trustify/blob/986f992cdf67c0504ccd52f1a0cd7d06f95cad14/common/src/db/func.rs#L51-L58

**Comment by bxf12315** [comment]
> > First, I would like to consider how to reuse these three methods: build_spdx_license_query, build_cyclonedx_license_query, and create_license_filtering_base_query. Second, I have done some testing, and the code is [here](https://github.com/mrizzi/trustify/pull/9). Currently, the search conditions are all for equality matches—I hope to add support for containment (i.e., 'contains') conditions. Third, I measured the execution performance in my tests, and it basically always takes more than 6 minutes.
> 
> 1. Re the reuse, I think it's an important goal but this PR introduces those functions for the first time. Once merged, it will be up to next changes to reuse them if and how it will be possible. For the time being, it helps readability, avoiding never ending methods
> 2. In [239059f](https://github.com/guacsec/trustify/commit/239059f6fde89e81a2428340530ba1541d3b74b0) I've added a test with the `~` condition to prove it works as well
> 3. We know the `search_purls` scale test has unreliable results. We can work in analyzing the query execution plan and optimize it to ensure better performances. The `SbomIdIdx` index creation is meant to cover this: it supports the execution of
> 
> ```
>         SELECT license_id, name
>         FROM licensing_infos
>         WHERE sbom_id = sbom_id_param
> ```
> 
> which is the only query executed in the `expand_license_expression` PLSQL function.

ok Thanks

**Review by bxf12315 — APPROVED** [review]

File: modules/fundamental/src/purl/service/mod.rs (hunk 1 above)
Comment (mrizzi) [reply modules/fundamental/src/purl/service/mod.rs:343]: @bxf12315 thanks for the idea about using `EXISTS` and I gave it a try, applying these changes to the code:
```rust
iff --git a/modules/fundamental/src/purl/service/mod.rs b/modules/fundamental/src/purl/service/mod.rs
index a390f71f..aaf66002 100644
--- a/modules/fundamental/src/purl/service/mod.rs
+++ b/modules/fundamental/src/purl/service/mod.rs
@@ -12,8 +12,8 @@ use sea_orm::{
     QueryOrder, QuerySelect, QueryTrait, RelationTrait, Select, prelude::Uuid,
 };
 use sea_query::{
-    ColumnType, Condition, Expr, Func, JoinType, Order, SelectStatement, SimpleExpr, UnionType,
-    extension::postgres::PgExpr,
+    BinOper, ColumnType, Condition, Expr, Func, JoinType, Order, SelectStatement, SimpleExpr,
+    UnionType, extension::postgres::PgExpr,
 };
 use tracing::instrument;
 use trustify_common::{
@@ -336,10 +336,7 @@ impl PurlService {
                 .union(UnionType::Distinct, select_packages_from_cyclonedx);
 
             select = select.filter(
-                Condition::all().add(
-                    qualified_purl::Column::Id
-                        .in_subquery(select_packages_filtering_by_license.clone()),
-                ),
+                Condition::all().add(Expr::exists(select_packages_filtering_by_license.take())),
             );
         }
 
@@ -406,6 +403,15 @@ impl PurlService {
                 JoinType::Join,
                 sbom_package_license::Relation::License.def(),
             )
+            .filter(
+                Expr::col((qualified_purl::Entity, qualified_purl::Column::Id)).binary(
+                    BinOper::Equal,
+                    Expr::col((
+                        sbom_package_purl_ref::Entity,
+                        sbom_package_purl_ref::Column::QualifiedPurlId,
+                    )),
+                ),
+            )
     }
 
     #[instrument(skip(self, connection), err)]
```
but then the `qualified_packages_filter_by_license` test execution slowed down a lot with a lot of warnings like
```bash
WARN purls{query=Query { q: "license=GPLv3+ with exceptions and GPLv2+ with exceptions and LGPLv2+ and BSD", sort: "" } paginated=Paginated { offset: 0, limit: 0 }}:fetch: sqlx::query: slow statement: execution time exceeded alert threshold summary="SELECT \"qualified_purl\".\"id\", \"qualified_purl\".\"versioned_purl_id\", \"qualified_purl\".\"qualifiers\", …" db.statement="\n\nSELECT \"qualified_purl\".\"id\", \"qualified_purl\".\"versioned_purl_id\", \"qualified_purl\".\"qualifiers\", \"qualified_purl\".\"purl\" FROM \"qualified_purl\" WHERE TRUE AND EXISTS(SELECT \"sbom_package_purl_ref\".\"qualified_purl_id\" FROM \"sbom_package_purl_ref\" JOIN \"sbom_package\" ON \"sbom_package_purl_ref\".\"sbom_id\" = \"sbom_package\".\"sbom_id\" AND \"sbom_package_purl_ref\".\"node_id\" = \"sbom_package\".\"node_id\" JOIN \"sbom_package_license\" ON \"sbom_package\".\"sbom_id\" = \"sbom_package_license\".\"sbom_id\" AND \"sbom_package\".\"node_id\" = \"sbom_package_license\".\"node_id\" JOIN \"license\" ON \"sbom_package_license\".\"license_id\" = \"license\".\"id\" WHERE \"qualified_purl\".\"id\" = \"sbom_package_purl_ref\".\"qualified_purl_id\" AND expand_license_expression(\"text\", \"sbom_package_license\".\"sbom_id\") = $1 AND (\"text\" ILIKE $2) UNION (SELECT \"sbom_package_purl_ref\".\"qualified_purl_id\" FROM \"sbom_package_purl_ref\" JOIN \"sbom_package\" ON \"sbom_package_purl_ref\".\"sbom_id\" = \"sbom_package\".\"sbom_id\" AND \"sbom_package_purl_ref\".\"node_id\" = \"sbom_package\".\"node_id\" JOIN \"sbom_package_license\" ON \"sbom_package\".\"sbom_id\" = \"sbom_package_license\".\"sbom_id\" AND \"sbom_package\".\"node_id\" = \"sbom_package_license\".\"node_id\" JOIN \"license\" ON \"sbom_package_license\".\"license_id\" = \"license\".\"id\" WHERE \"qualified_purl\".\"id\" = \"sbom_package_purl_ref\".\"qualified_purl_id\" AND \"license\".\"text\" = $3))\n" rows_affected=0 rows_returned=0 elapsed=3.772697958s elapsed_secs=3.772697958 slow_threshold=1s
```
so I didn't apply the change.

File: migration/src/m0001140_expand_spdx_licenses_function/expand_up.sql (hunk 2 above)
Comment (mrizzi) [reply migration/src/m0001140_expand_spdx_licenses_function/expand_up.sql]: I've added `STABLE PARALLEL SAFE` following what we've been doing for other functions:
- `STABLE` because it will return the same result within a single statement execution, but the result could change between different executions if the `licensing_infos` table data is modified
- `PARALLEL SAFE` because the function only reads from `licensing_infos` table, processes text locally and returns a result without any database modifications. All operations are safe for concurrent execution.

Comment (mrizzi) [reply migration/src/m0001140_expand_spdx_licenses_function/expand_up.sql]: @JimFuller-RedHat I'm moving forward with merging this PR because it blocks some other tasks but please, feel free to come back to me with any suggestion/comment you find and I'll enhance this in a separated PR.
